openstack/designate/bin/health-probe.py

```diff
@@ -0,0 +1,335 @@
+        client.call(context.RequestContext(), "oslo_rpc_server_ping", arg=None)
+    except oslo_messaging.exceptions.MessageDeliveryFailure:
+        # Log to pod events
+        sys.stderr.write("Health probe unable to reach message bus")
+        sys.exit(0)  # return success
+    except oslo_messaging.rpc.client.RemoteError as re:
+        message = getattr(re, "message", str(re))
+        if ("Endpoint does not support RPC method" in message) or ("Endpoint does not support RPC version" in message):
+            sys.exit(0)  # Call reached the service
+        else:
+            sys.stderr.write("Health probe unable to reach service")
+            sys.exit(1)  # return failure
+    except oslo_messaging.exceptions.MessagingTimeout:
+        sys.stderr.write("Health probe timed out. Agent is down or response timed out")
+        sys.exit(1)  # return failure
+    except Exception as ex:
+        message = getattr(ex, "message", str(ex))
+        sys.stderr.write("Health probe caught exception sending message to service: %s" % message)
+        sys.exit(0)
+    except:
+        sys.stderr.write("Health probe caught exception sending message to service")
+        sys.exit(0)
+    finally:
+        if client.transport:
+            client.transport.cleanup()
+
+
+def _tcp_reachable(host: str, port: int, timeout: float = 5) -> bool:
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.settimeout(timeout)
+    try:
+        sock.connect((host, port))
+        return True
+    finally:
+        sock.close()
+
+
+@retry(stop=stop_after_attempt(3), wait=wait_fixed(1))
+def tcp_reachable(host: str, port: int) -> bool:
+    """Check if TCP socket is reachable. Retries 3 times before giving up"""
+    try:
+        return _tcp_reachable(host, port)
+    except Exception:
+        return False
+
+
+def tcp_socket_status(process: Optional[str], ports: Set[int]) -> int:
+    """Check the TCP socket status on a process"""
+    for p in psutil.process_iter():
+        try:
+            with p.oneshot():
+                if process in " ".join(p.cmdline()):
+                    pcon = p.connections()
+                    for con in pcon:
+                        try:
+                            rport = con.raddr[1]
+                            status = con.status
+                        except IndexError:
+                            continue
+                        if rport in ports and status == tcp_established:
+                            return 1
+        except psutil.Error:
+            continue
+    return 0
+
+
+def tcp_connection_established(ports: Set[int]) -> int:
+    """Check the presence of established TCP connections to remote ports"""
+    scon = psutil.net_connections(kind="tcp")
+    for con in scon:
+        try:
+            rport = con.raddr[1]
+            status = con.status
+        except IndexError:
+            continue
```

Comment:
We are interested only in established connections.  `tcp_connection_established` function returns non-zero only when no such connections exist.

If the connection is not established, then `con.raddr` is an empty tuple. The same is true for connections in status `LISTEN`, e.g.:

`sconn(fd=-1, family=<AddressFamily.AF_INET6: 10>, type=<SocketKind.SOCK_STREAM: 1>, laddr=addr(ip='::', port=5778), raddr=(), status='LISTEN', pid=None),`

https://psutil.readthedocs.io/en/stable/#psutil.Process.connections
>raddr: the remote address as a (ip, port) named tuple or an absolute path in case of UNIX sockets. When the remote endpoint is not connected you’ll get an empty tuple (AF_INET*) or "" (AF_UNIX). For UNIX sockets see notes below.